### PR TITLE
[JAX] Generalizing Activation Primitives

### PR DESCRIPTION
**Pull request by phu0ngng**
This PR generalized the primitive blocks and respective C++ functions for activations.
Note that inputs for all activation functions should have `x.shape[-2] == len(activation_type)`.
Tests and LayerNormMLP were also updated accordingly.

### CODE REVIEW

File: transformer_engine/jax/csrc/modules.cpp [hunk 1]
```diff
@@ -211,107 +235,93 @@ void GeluFP8(cudaStream_t stream, void **buffers, const char *opaque, size_t opa
...
+        nvte_dswish(input_tensor.data(), act_input_tensor.data(),
+                    output_tensor.data(), stream);
+        break;
+      case ActivationEnum::SWIGLU:
+        nvte_dswiglu(input_tensor.data(), act_input_tensor.data(),
+                     output_tensor.data(), stream);
+        break;
+      default:
+        // TODO: what is a good message here
+        throw std::runtime_error("Not Implemented");
```
Comment (zlsh80826) [inline transformer_engine/jax/csrc/modules.cpp:283]: There is `NVTE_ERROR` marco for raising error for unexpected behavior.
For example, you can use it here as
```
NVTE_ERROR("Unsupported ActivationEnum");
```

File: transformer_engine/jax/cpp_extensions.py [hunk 2]
```diff
@@ -124,6 +124,15 @@ def _check_valid_batch_dims(bdims):
             f"but got {dim=}"
 
 
+# The value of activations are explicitedly defined to avoid order rearrangement in the future
+ActivationEnum = {
+    ('gelu',): 0,
+    ('gelu', 'linear'): 1,
+    ('silu',): 2,
+    ('silu', 'linear'): 3
+}
```
Comment (zlsh80826) [inline transformer_engine/jax/cpp_extensions.py:133]: I prefer to expose the C++ ActivationEnum to here through extensions.cpp, this can generate a hard dependency between python/c++'s enum to prevent the misalignments for the future development. You can refer `NVTE_QKV_Layout`.

File: tests/jax/test_custom_call_compute.py [hunk 3]
```diff
@@ -204,8 +204,8 @@ def test_grad_fused_layernorm_fp8_mlp(self, m, n, k,
             b1 = jax.random.normal(subkeys[3], (len(activation_type), n), jnp.bfloat16)
             b2 = jax.random.normal(subkeys[4], (k,), jnp.bfloat16)
         else:
-            b1 = jax.random.normal(subkeys[3], (0,), jnp.bfloat16)
-            b2 = jax.random.normal(subkeys[4], (0,), jnp.bfloat16)
+            b1 = jnp.empty((0,), jnp.bfloat16)
+            b2 = jnp.empty((0,), jnp.bfloat16)
```
Comment (zlsh80826) [inline tests/jax/test_custom_call_compute.py:208]: Does `None` work if there is no_bias? I think `None` may be more used in the real case

File: transformer_engine/jax/cpp_extensions.py [hunk 4]
```diff
@@ -4356,13 +4181,14 @@ def abstract(dz_aval, amax_aval, scale_aval, scale_inv_aval, *, out_dtype,
         assert amax_aval.dtype == jnp.float32
         assert scale_aval.dtype == jnp.float32
         assert scale_inv_aval.dtype == jnp.float32
-        gi_hidden_size = dz_aval.shape[-1]
+        if dz_aval.shape[-2] == 1 or dz_aval.shape[-2] == 2:
+            gi_hidden_size = reduce(operator.mul, dz_aval.shape[-2:])
+        else:
+            gi_hidden_size = dz_aval.shape[-1]
```
Comment (zlsh80826) [inline transformer_engine/jax/cpp_extensions.py:4187]: When will dz_aval.shape[-2] != 1 or 2?

File: transformer_engine/jax/cpp_extensions.py [hunk 5]
```diff
@@ -5993,36 +4528,41 @@ def partition(out_dtype, static_axis_boundary, mesh, arg_infos, result_infos):
...
+    dz: jnp.ndarray, x: jnp.ndarray, amax: jnp.ndarray, scale: jnp.ndarray,
+    scale_inv: jnp.ndarray, out_dtype: TEDType,
+    static_axis_boundary: int,
+    activation_type: Sequence[Union[str, Callable]]
+    ) -> Tuple[jnp.ndarray, jnp.ndarray, jnp.ndarray]:
     """
-    cast transpose d_gated_silu fusion wrapper
-    Return FP8(dgeglu(inputs))
+    cast transpose d_gated_act_lu fusion wrapper
+    Return FP8(dgayed_act_lu(inputs))
```
Comment (zlsh80826) [inline transformer_engine/jax/cpp_extensions.py:4557]: typo: dgated_act_lu

File: transformer_engine/jax/csrc/modules.h [hunk 6]
```diff
@@ -43,26 +43,37 @@ struct Shape {
     }
 };
 
+enum ActivationEnum {
```
Comment (zlsh80826) [inline transformer_engine/jax/csrc/modules.h:46]: Use `enum class` instead of `enum`.

File: transformer_engine/jax/mlp.py [hunk 7]
```diff
@@ -253,8 +224,8 @@ def _fused_layernorm_fp8_mlp_fwd_rule(
 
     # Squeeze act axis
     # (hidden_in, 1, hidden_out) -> (hidden_in, hidden_out)
-    if not is_gated:
-        kernel_1 = jnp.squeeze(kernel_1, axis=-2)
+#    if not is_gated:
```
Comment (zlsh80826) [inline transformer_engine/jax/mlp.py:227]: Remove the unnecessary comments.

File: transformer_engine/jax/csrc/modules.cpp [hunk 8]
```diff
@@ -170,31 +173,52 @@ void CastTranspose(cudaStream_t stream, void **buffers, const char *opaque, size
                         input_cast_trans_tensor.data(), stream);
 }
 
-void GeluImpl(void *input, size_t m, size_t n, DType in_dtype, DType out_dtype, float *scale,
-              cudaStream_t stream, float *scale_inverse, float *amax, void *output) {
-    auto input_shape = std::vector<size_t>{m, n};
+void ActLuImpl(void *input, size_t m, size_t n, DType in_dtype, DType out_dtype, float *scale,
+              cudaStream_t stream, float *scale_inverse, float *amax, void *output,
+              size_t act_enum) {
```
[the remark this answers is not in the document]
Comment (phu0ngng) [reply transformer_engine/jax/csrc/modules.cpp:178]: Good idea. Agree!

File: tests/jax/test_custom_call_compute.py (hunk 3 above)
Comment (phu0ngng) [reply tests/jax/test_custom_call_compute.py:208]: Hi, as we have a few `reshape` based on `bias.shape` inside the MLP, `None` would break the code.
I would keep those as they are.

File: transformer_engine/jax/cpp_extensions.py (hunk 4 above)
Comment (phu0ngng) [reply transformer_engine/jax/cpp_extensions.py:4187]: Hi, the first `dbias_cast_transpose` in the `_fused_layernorm_fp8_mlp_bwd` is with `grad` input which has the size of `[batch_size, out]`. So its `shape[-2]` is neither 1 nor 2.

File: transformer_engine/jax/csrc/modules.cpp [hunk 9]
```diff
@@ -37,6 +37,19 @@ std::vector<size_t> MakeShapeVector(NVTEShape shape) {
...
+    case NVTE_Activation_Enum::GEGLU: return 2;
+    case NVTE_Activation_Enum::SILU: return 1;
+    case NVTE_Activation_Enum::SWIGLU: return 2;
+    default:
+      NVTE_ERROR("Unsupported ActivationEnum");
+      break;
+    return -1;
+  }
+}
+
```
Comment (phu0ngng) [inline transformer_engine/jax/csrc/modules.cpp:52]: @zlsh80826 Do you have a better way to extract `act_len` here?

Comment (zlsh80826) [reply transformer_engine/jax/csrc/modules.cpp:52]: This looks good to me!

File: transformer_engine/jax/cpp_extensions.py (hunk 4 above)
Comment (zlsh80826) [reply transformer_engine/jax/cpp_extensions.py:4186]: For this cases, the code here is not safe. It is possible that the batch_size = 2, then the gi_hidden_size will be computed incorrectly. Can you try to use other metrics to determine if the input is gated? (I'm not sure but probably the transpose_axis_boundary)

File: tests/jax/test_custom_call_compute.py (hunk 3 above)
Comment (zlsh80826) [reply tests/jax/test_custom_call_compute.py:208]: We can try this but we need to aware that if there is a performance regression with 0 shape bias during the e2e training. I afriad that there are additional all-2-all or grad computations.

Comment (denera) [reply tests/jax/test_custom_call_compute.py:208]: @zlsh80826 I'm very surprised about this regression.

Why/how does JAX/XLA trigger any extra compute for ops that we have written our custom FWD/BWD implementations? We have full control over what happens in these ops, so I would expect any extra unnecessary compute with an empty bias to be our own bug to fix in TE.

I also don't quite understand how this can trigger extra communication. Our user-facing interfaces for ops/layers with optional biases are written to accept `bias = None` arguments. We internally replace these `None` terms with an *unsharded* `jnp.empty(0)` tensor before calling the underlying custom ops. This should exactly match the `bias_sharding = NamedSharding(mesh, PartitionSpec(None))` we enforce in the primitives' `partition(...)` and `infer_sharding_from_operands(...)` calls, and XLA should not have any reason to trigger extra communication. If it somehow still does trigger extra comms, that suggests either incorrect use of custom partitioning on our end, or incorrect parsing of the primitives in JAX/XLA.

How hard would it be to collect some hard evidence for any extra compute/comm that these empty biases may be triggering in e2e training? It would be great to see an execution profile for a single FWD/BWD pass so we can figure out what (if anything) we need to do in TE custom ops to eliminate them, or otherwise pass the issue over to the JAX/XLA team if it's their bug to fix.

File: transformer_engine/jax/cpp_extensions.py (hunk 4 above)
Comment (phu0ngng) [reply transformer_engine/jax/cpp_extensions.py]: I pushed the change in which transpose_axis_boundary is used for conditional check. 
I left a note and added an error raise there, in case if it causes any issue in the future.

File: tests/jax/test_custom_call_compute.py (hunk 3 above)
Comment (denera) [reply tests/jax/test_custom_call_compute.py:208]: Is there any reason why we can't support `biases = [ None, None ]` for `_fused_layernorm_fp8_mlp(...)`?

It should be trivial to replace the check on `use_bias` with checks on `bias_1 is not None` or `bias_2 is not None` during the forward pass, and then save `use_bias_1` and `use_bias_2` flags into the autodiff context so we know what gradients to compute in the backward pass.

**Review by denera — CHANGES_REQUESTED** [review]

File: tests/jax/test_custom_call_compute.py [hunk 10]
```diff
@@ -310,19 +310,19 @@ def ref_func(x, ln_s, y, z, w, v, fp8_max, fp8_metas_amax, fp8_metas_scale,
...
         if use_bias:
-            assert_allclose(jnp.asarray(primitive_b1_grad, np.float32),
-                            jnp.asarray(ref_b1_grad, np.float32),
-                            dtype=jnp.bfloat16)
             assert_allclose(jnp.asarray(primitive_b2_grad, np.float32),
                             jnp.asarray(ref_b2_grad, np.float32),
                             dtype=jnp.bfloat16)
+            assert_allclose(jnp.asarray(primitive_b1_grad, np.float32),
+                            jnp.asarray(ref_b1_grad, np.float32),
+                            dtype=jnp.bfloat16)
```
[the remark this answers is not in the document]
Comment (phu0ngng) [reply tests/jax/test_custom_call_compute.py]: The tests both passed with `BWD_DTYPE`.

File: tests/jax/test_custom_call_compute.py (hunk 3 above)
Comment (phu0ngng) [reply tests/jax/test_custom_call_compute.py]: So, I adjusted the `fused_layernorm_fp8_mlp` so that it can take either `None` or `jnp.empty` as values for biases, providing that the `use_bias = False`.
Could you have a look and let me know if there is an issue?

Comment (zlsh80826) [reply tests/jax/test_custom_call_compute.py]: I discovered another issue with zero-shape bias for no-bias cases while resolving the conflict in [PR-796](https://github.com/NVIDIA/TransformerEngine/pull/796). [PR-765](https://github.com/NVIDIA/TransformerEngine/pull/765) began using [zero-shape bias](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/flax/module.py#L1038-L1052) in the MLP module, and I observe that zero-shape bias is created in the Flax model parameters. I believe we should avoid this, as it can confuse users about why biases appear in the model parameters when they did not specify their use. This could also cause the weight converter to fail because the non-TE checkpoint does not include the bias. In [PR-796](https://github.com/NVIDIA/TransformerEngine/pull/796), I had a [WAR](https://github.com/NVIDIA/TransformerEngine/pull/796/commits/5be3f4b1b56756e1db7e435b761fbe733ea32ab6) to remove the reduntant bias in the model parameters. As per @phu0ngng's latest change, the `fused_layernorm_fp8_mlp` can now accept `[None, None]` for no-bias cases. @phu0ngng could you help revise [jnp.empty(0)](https://github.com/NVIDIA/TransformerEngine/pull/796/commits/5be3f4b1b56756e1db7e435b761fbe733ea32ab6#diff-a39a0038aabb6a98b69edbaf40891bf61ec8a77909f175f60128c29b8bbbb1aaR1055-R1056) to `None` after PR-796 merged?

File: transformer_engine/jax/cpp_extensions.py [hunk 11]
```diff
@@ -4356,13 +4180,17 @@ def abstract(dz_aval, amax_aval, scale_aval, scale_inv_aval, *, out_dtype,
...
         assert scale_aval.dtype == jnp.float32
         assert scale_inv_aval.dtype == jnp.float32
-        gi_hidden_size = dz_aval.shape[-1]
+        # Phuong: input shape [batch_size, n_out] should have transpose_axis_boundary = -1
+        if transpose_axis_boundary == -2:
+            gi_hidden_size = reduce(operator.mul, dz_aval.shape[-2:])
+        elif transpose_axis_boundary == -1:
+            gi_hidden_size = dz_aval.shape[-1]
+        else:
+            raise NotImplementedError
```
Comment (zlsh80826) [inline transformer_engine/jax/cpp_extensions.py:4189]:
```suggestion
     gi_hidden_size = reduce(operator.mul, dz_aval.shape[transpose_axis_boundary:])
```

**Review by denera — COMMENTED** [review]
LGTM! We can merge once @zlsh80826 also approves for his requested changes and CI comes clean.

**Review by denera — APPROVED** [review]

File: tests/jax/test_custom_call_compute.py (hunk 3 above)
Comment (phu0ngng) [reply tests/jax/test_custom_call_compute.py]: I set `bias = None` in the flax modules.

**Review by zlsh80826 — APPROVED** [review]
LGTM!
